hdf5 1.8.22 parallel Fortran interface is broken with CMake

### Steps to reproduce

I install hdf5 1.8.22 with
```
bin/spack install  hdf5@1.8.22%gcc+mpi+szip+fortran+cxx+hl ^openmpi schedulers=slurm +pmi +thread_multiple +legacylaunchers fabrics=ucx ^ucx+mlx5_dv
```
and then loaded openmpi
```
bin/spack module tcl refresh -y
module use -p share/spack/modules/linux-sles15-haswell/ 
module load openmpi-1.10.7-gcc-11.2.0-5p475pq
```
and tried to compile the hdf5 fortran example [ph5example.f90](https://raw.githubusercontent.com/HDFGroup/hdf5/develop/fortran/examples/ph5example.f90
)
via
```
mpifort ph5example.f90 -Iopt/spack/linux-sles15-haswell/gcc-11.2.0/hdf5-1.8.22-eqz3y6v6ibvvirocnnphhq6aa63czcwd/include -Lopt/spack/linux-sles15-haswell/gcc-11.2.0/hdf5-1.8.22-eqz3y6v6ibvvirocnnphhq6aa63czcwd/lib -lhdf5_fortran -lhdf5
```

### Error message

```
/usr/bin/ld: /tmp/ccV3VZVL.o: in function `MAIN__':
ph5example.f90:(.text+0x33c): undefined reference to `h5pset_fapl_mpio_f_'
/usr/bin/ld: ph5example.f90:(.text+0x6ae): undefined reference to `h5pset_dxpl_mpio_f_'
collect2: error: ld returned 1 exit status
```

If I use the old package.py with the autotools, it builds and compiles correctly. I also built hdf5 by hand with cmake and there the symbol is also missing. I also checked with intel-mpi/intel-compilers, no difference. For newer versions of hdf5 this problem does not exist.

### Information on your system

* **Spack:** 0.19.0.dev0 (cfc46504ac64049df98e5876396a59ccf3deebf9)
* **Python:** 3.6.12
* **Platform:** linux-sles15-haswell
* **Concretizer:** clingo

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have searched the issues of this repo and believe this is not a duplicate
- [X] I have run the failing commands in debug mode and reported the output

--- a/var/spack/repos/builtin/packages/hdf5/package.py
+++ b/var/spack/repos/builtin/packages/hdf5/package.py
@@ -111,6 +111,8 @@ class Hdf5(CMakePackage):
     conflicts('+java', when='~shared')
     # Fortran fails built with shared for old HDF5 versions
     conflicts('+fortran', when='+shared@:1.8.15')
+    # See https://github.com/spack/spack/issues/31085
+    conflicts('+fortran+mpi', when='@1.8.22')
 
     # There are several officially unsupported combinations of the features:
     # 1. Thread safety is not guaranteed via high-level C-API but in some cases

--- a/var/spack/repos/builtin/packages/hdf5/package.py
+++ b/var/spack/repos/builtin/packages/hdf5/package.py
@@ -73,7 +73,8 @@ class Hdf5(CMakePackage):
     variant('hl', default=False, description='Enable the high-level library')
     variant('cxx', default=False, description='Enable C++ support')
     variant('fortran', default=False, description='Enable Fortran support')
-    variant('java', default=False, description='Enable Java support')
+    variant('java', when='@1.10:', default=False,
+            description='Enable Java support')
     variant('threadsafe', default=False,
             description='Enable thread-safe capabilities')
     variant('tools', default=True, description='Enable building tools')
@@ -104,9 +105,6 @@ class Hdf5(CMakePackage):
     conflicts('api=v18', when='@1.6.0:1.6',
               msg='v18 is not compatible with this release')
 
-    # The Java wrappers and associated libhdf5_java library
-    # were first available in 1.10
-    conflicts('+java', when='@:1.9')
     # The Java wrappers cannot be built without shared libs.
     conflicts('+java', when='~shared')
     # Fortran fails built with shared for old HDF5 versions

